plugins/inputs/gnmi/handler.go

```diff
@@ -231,30 +230,33 @@ func (h *handler) handleSubscribeResponseUpdate(acc telegraf.Accumulator, respon
 				h.emptyNameWarnShown = true
 			}
 		}
+		aliasInfo := newInfoFromString(aliasPath)
 
 		// Group metrics
-		fieldPath := field.path.String()
-		key := strings.ReplaceAll(fieldPath, "-", "_")
+		var key string
 		if h.canonicalFieldNames {
 			// Strip the origin is any for the field names
-			if parts := strings.SplitN(key, ":", 2); len(parts) == 2 {
+			if parts := strings.SplitN(strings.ReplaceAll(field.path.String(), "-", "_"), ":", 2); len(parts) == 2 {
 				key = parts[1]
 			}
 		} else {
-			if len(aliasPath) < len(key) && len(aliasPath) != 0 {
-				// This may not be an exact prefix, due to naming style
-				// conversion on the key.
-				key = key[len(aliasPath)+1:]
-			} else if len(aliasPath) >= len(key) {
+			// If the origins are the same and the alias is shorter than the
+			// full path to avoid an empty key, then strip the common part if
+			// the field is prefixed with the alias path.
+			if aliasInfo.origin == field.path.origin && aliasInfo.isSubPathOf(field.path) && len(aliasInfo.segments) < len(field.path.segments) {
```

Comment:
I think you don't need to check the segment length as it is done in `isSubPathOf`, isn't it? Furthermore, the `origin` check is wrong as `foo:/some/path` is the same as `/some/path` but not the same as `bar:/some/path`... That's why I suggested the use of `isSubPathOf` in the first place...

Reply:
PR updated

>  I think you don't need to check the segment length as it is done in isSubPathOf, isn't it?

As I mention in the inline docs, it is to avoid an empty key due to them being the same length. The `full path field key` test is an example.

>  Furthermore, the origin check is wrong as foo:/some/path is the same as /some/path

I will added this to the inline docs as this is entirely not obvious. I had assumed incorrectly that the origin was like a namespace.

Reply:
> I had assumed incorrectly that the origin was like a namespace.

It is kind of but the spec says that it _might_ be there if there is no collision...